File: backend/openmtscied/shared/crawlers/khan_crawler.py
"""
Khan Academy 爬虫包装器
"""

from . import register_crawler
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent / "scripts" / "scrapers"))


def crawl_khan_academy_wrapper(config: dict) -> dict:
    """Khan Academy 爬虫包装器"""
    try:
        from generate_khan_academy import generate_khan_academy_courses
        
        courses = generate_khan_academy_courses()
        
        output_file = config.get('output_file', 'data/course_library/khan_academy_courses.json')
        Path(output_file).parent.mkdir(parents=True, exist_ok=True)
        
        import json
        with open(output_file, 'w', encoding='utf-8') as f:
            json.dump(courses, f, ensure_ascii=False, indent=2)
        
        return {
            "success": True,
            "total_items": len(courses),
            "scraped_items": len(courses),
            "data": courses
        }
    except Exception as e:
        return {
            "success": False,
            "total_items": 0,
            "scraped_items": 0,
            "error": str(e),
            "data": []
        }


register_crawler(
    crawler_id="khan_academy",
    name="Khan Academy Courses",
    handler=crawl_khan_academy_wrapper,
    description="生成可汗学院 K-12 STEM 课程"
)

File: backend-next/app/api/v1/admin/crawler/lib/khan-crawler.ts
/**
 * Khan Academy 爬虫
 * 从现有数据文件加载可汗学院课程
 */

import { registerCrawler } from './registry';
import type { CrawlerConfig, CrawlResult } from './types';
import { loadJsonFile, createSuccessResult, createErrorResult } from './utils';

async function crawlKhanAcademy(config: CrawlerConfig): Promise<CrawlResult> {
  try {
    console.log('[Khan Academy] Starting crawler...');
    
    const outputFile = config.output_file || 'course_library/khan_academy_courses.json';
    
    // 加载现有数据
    const data = loadJsonFile(outputFile);
    
    if (data.length === 0) {
      console.warn(`[Khan Academy] No data found in ${outputFile}`);
      return createSuccessResult(0, 0, []);
    }
    
    console.log(`[Khan Academy] Loaded ${data.length} courses from ${outputFile}`);
    
    return createSuccessResult(data.length, data.length, data);
  } catch (error: unknown) {
    console.error('[Khan Academy] Crawler error:', error);
    return createErrorResult(error.message || 'Unknown error');
  }
}

// 注册爬虫
registerCrawler(
  'khan_academy',
  'Khan Academy Courses',
  crawlKhanAcademy,
  '生成可汗学院 K-12 STEM 课程'
);

console.log('[Khan Academy] Crawler registered');
